### PR TITLE
Add a demo with a javascript library

### PR DESCRIPTION
**Pull request by arn-the-long-beard**
Well, it would be nice to have demo to see how we can integrate the use of Javascript libraries into MZ.

In this context I succeed to get rendering with a rich text with `Quill.js` but if you try the example you will see that we also get the Javascript rendered as `text` while being correctly understood by the browser.

Not sure how to deal with that.

Here is the problematic code I have: 

```rust

fn root() -> impl Element {
    Column::new()
        .item(
            El::new()
                .update_raw_el(|el| el.attr("class", "standalone-container"))
                .child(
                    El::new().update_raw_el(|el| el.attr("id", "snow-container")),
                ),
        )
        .item(El::with_tag(Tag::Custom("script")).child(
            r#"
  var quill = new Quill('#snow-container', {
    placeholder: 'Compose an epic...',
    theme: 'snow'
  });
"#,
        ))
}


```
If I move everything into the `html` String served by the backend, then I have no issue. But I think the `js` part should be moved to front end since there is a need to interact with it.

image

### CODE REVIEW

**Comment by arn-the-long-beard** [comment]
Okay, the js code was displaying because of some styling that I updated.

**Comment by MartinKavik** [comment]
(Updates in https://github.com/arn-the-long-beard/MoonZoon/pull/1)

**Comment by arn-the-long-beard** [comment]
As written in the PR comments on my fork => Thanks a lot for the help :heart: .
I went into the commits and they do make sens a lot. I have no idea for improving what we have right now. 
So nice to see the component we have at the end :grin:

**Comment by arn-the-long-beard** [comment]
I made a `git rebase` instead of cherry pick way I usually do so the historic is conserved correctly.

**Comment by MartinKavik** [comment]
Please rebase on `main` to resolve conflcts.
Can I merge it then?

**Comment by arn-the-long-beard** [comment]
Well, I understand what this code is doing ( loading the js+css files ) : 

```rust

#[static_ref]
fn load_assets_once() -> &'static Mutable<bool> {
    Task::start(async {
        join!(
            load_stylesheet("https://cdn.quilljs.com/1.3.6/quill.snow.css"),
            load_script("https://cdn.quilljs.com/1.3.6/quill.min.js"),
        );
        load_assets_once().set(true);
    });
    Mutable::default()
}


```
But I might not understand how it does it actually. I am a bit confused with `load_assets_once().set(true) ` 

When we call the function `load_assets_once()` , we do spawn the `Task`, but it will triggers only when we do change the value on the `Mutable`, that is why we need to use `set(true)` ?

I went into the `futures_signals` documentation about `Mutable` to try to understand. Can you complete my understanding ?

**Comment by arn-the-long-beard** [comment]
> Please rebase on `main` to resolve conflcts. Can I merge it then?

The rebease isn't correct ? Everything seems Ok on my side from the `rebase` I made. If lights are green on yours, then you can merge :smile:

File: examples/js_text_editor/frontend/src/text_editor.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    #[inline]
+    fn into_iter(self) -> Self::IntoIter {
+        iter::once(self)
+    }
+}
+
+// ------ ------
+//    Javascript Snippet Integration
+// ------ ------
```
Comment (MartinKavik) [inline examples/js_text_editor/frontend/src/text_editor.rs:87]: This is pretty ugly :crying_cat_face:

Comment (arn-the-long-beard) [reply examples/js_text_editor/frontend/src/text_editor.rs]: I made a tiny change.

**Comment by arn-the-long-beard** [comment]
I really do not know why we have conflict. When I do the rebase on my machine, I got no issue. I can do it with the cherry pick the regular way, but as written there when we have different branch , it is no the best option, because your PR might disappear from the history. 
https://stackoverflow.com/questions/9229301/git-cherry-pick-says-38c74d-is-a-merge-but-no-m-option-was-given

**Comment by MartinKavik** [comment]
I've hard-reset the branch to the commit just before the merge commit, added a new commit (I've chosen `js_bridge` instead of `js_snippet` because you can find multiple `bridge` modules in Rust/Wasm apps) and then I've forced-pushed the branch. The problem was the merge commit / reference to other branch I think. I don't use cherry picks or other branch manipulation staff when I don't have to - `commit` + `rebase` + `push` is everything I need. Please do a finally look and I'll merge it, thanks!

**Comment by MartinKavik** [comment]
> Well, I understand what this code is doing ( loading the js+css files ) :
> 
> ```rust
> #[static_ref]
> fn load_assets_once() -> &'static Mutable<bool> {
>     Task::start(async {
>         join!(
>             load_stylesheet("https://cdn.quilljs.com/1.3.6/quill.snow.css"),
>             load_script("https://cdn.quilljs.com/1.3.6/quill.min.js"),
>         );
>         load_assets_once().set(true);
>     });
>     Mutable::default()
> }
> ```
> 
> But I might not understand how it does it actually. I am a bit confused with `load_assets_once().set(true) `
> 
> When we call the function `load_assets_once()` , we do spawn the `Task`, but it will triggers only when we do change the value on the `Mutable`, that is why we need to use `set(true)` ?
> 
> I went into the `futures_signals` documentation about `Mutable` to try to understand. Can you complete my understanding ?

`load_assets_once`'s body is executed only once - on the first function call. Following calls only return the `&'static Something` immediately.
`Task::start` just spawns and leaks a `Future`. It's a non-blocking call. The `Future` is resolved in the future - until then, `load_assets_once().signal()` returns always `false` when polled.
After both `load_*` calls are concurrently awaited by `join!`, the `Mutable` is toggled to `true`. It means the resulted `signal` is switched and always returns `true` from now.

The same code could be rewritten to something like this:
```rust
#[static_ref]
fn load_assets_once() -> &'static Mutable<bool> {
    let loaded = Mutable::new(false);
    let loaded_for_task = loaded.clone();
 
    Task::start(async move {
        join!(
            load_stylesheet("https://cdn.quilljs.com/1.3.6/quill.snow.css"),
            load_script("https://cdn.quilljs.com/1.3.6/quill.min.js"),
        );
        loaded_for_task.set(true);
    });

    loaded
}
```

**Comment by arn-the-long-beard** [comment]
`js_bdridge` is very good.  I should have do like you do and comment on the commit rather than making one to change your code ( I hope it was not rude ). Next time I do that to propose a change :wink: 

`commit + rebase + push `, yes this is what I did, maybe I messed up something somewhere. But you fixed it now :grin: 

Everything is alright now.

Thank you for the explanation for the `load_assets_once()`

You can merge.

**Pull request merged**